### PR TITLE
Add bpf() API

### LINKED ISSUE #559: Add shim for bpf syscall so other go libraries can be ported to windows
From [ebpf-for-windows slack channel](https://app.slack.com/client/T1MATJ4SZ/C0210QTK2MV/thread/C0210QTK2MV-1630774655.005900):

**Lorenz Bauer:** Have you decided what your ABI / API boundary is going to be? Is it raw syscalls or the libbpf C API?

**gmassaro:** is there an issue tracking the syscalls support ?

**Dave Thaler:**  `@Lorenz Bauer` if I understand your question, it's currently the libbpf api.  syscalls themselves work differently on windows vs linux.  We may be able to shim linux-like syscalls for things (like the go library?) that do syscalls directly without libbpf, but haven't done any design work on that yet.  Let us know if that's what you'd like.

**Lorenz Bauer:** yeah that was my question, thanks :slightly_smiling_face:

**Dave Thaler:** sure, are you interested in something in particular that makes raw syscalls?   I'd like to file a github issue to track this, so more information would be helpful.

**Dave Thaler:** or even better, you're welcome to file it :)

**Lorenz Bauer:** more of an observation: Go packages tend to favour raw syscalls since they are faster than C FFI aka CGo and avoiding CGo makes distribution easier (no need for a C compiler). this is most important to read / write maps i'd say.

**Zvi Effron:** Plus, disabling CGo entirely makes things easier when writing code in the same repository that is C but not go related, because otherwise the Go compiler will complain about the C files. Especially when cross-compiling (i.e. my team develops on macOS, but we compile for Linux, or, if we started using eBPF for Windows, Windows.). (edited)

### PR DESCRIPTION
**Pull request by dthaler**
Fixes #559

This PR also adds a test that enumerates keys in a map (previously we had no such test for bpf_map_get_next_key()).

### CODE REVIEW

File: include/linux/bpf.h [hunk 1]
```diff
@@ -29,3 +30,131 @@ enum bpf_stats_type
...
+    BPF_LINK_GET_FD_BY_ID,
+    BPF_OBJ_GET_INFO_BY_FD,
+    BPF_LINK_DETACH,
+    BPF_PROG_BIND_MAP,
+};
+
+/// Attributes used by BPF_OBJ_GET_INFO_BY_FD.
+typedef struct
+{
+    uint32_t bpf_fd;   ///< File descriptor referring to an eBPF object.
```
Comment (poornagmsft) [inline include/linux/bpf.h:58]: nit: is bpf_ required ?

Comment (poornagmsft) [reply include/linux/bpf.h:58]: if it is along below of link_, prog_, would suggest a specific prefix be used

**Review by poornagmsft — APPROVED** [review]

File: include/linux/bpf.h (hunk 1 above)
Comment (dthaler) [reply include/linux/bpf.h:58]: The field names must match what Linux etc has so code can be portable.

File: tests/unit/libbpf_test.cpp [hunk 2]
```diff
@@ -1204,3 +1204,148 @@ TEST_CASE("bpf_object__load", "[libbpf]")
     REQUIRE(bpf_link__destroy(link) == 0);
     bpf_object__close(object);
 }
+
+// Test bpf() with the following command ids:
+// BPF_PROG_LOAD, BPF_OBJ_GET_INFO_BY_FD, BPF_PROG_GET_NEXT_ID,
+// BPF_MAP_CREATE, BPF_MAP_GET_NEXT_ID, BPF_PROG_BIND_MAP,
+// and BPF_MAP_GET_FD_BY_ID.
+TEST_CASE("BPF_PROG_BIND_MAP", "[libbpf]")
```
Comment (saxena-anurag) [inline tests/unit/libbpf_test.cpp:1212]: Minor: should we rename the test case name, as it looks like right now that it is only testing `BPF_PROG_BIND_MAP`, but it is actually testing many other IDs.

File: tests/unit/libbpf_test.cpp [hunk 3]
```diff
@@ -1204,3 +1204,148 @@ TEST_CASE("bpf_object__load", "[libbpf]")
...
+
+    // Release our own references on the map and program.
+    Platform::_close(map_fd);
+    Platform::_close(program_fd);
+}
+
+// Test bpf() with the following command ids:
+// BPF_MAP_CREATE, BPF_MAP_UPDATE_ELEM, BPF_MAP_LOOKUP_ELEM,
+// BPF_MAP_GET_NEXT_KEY, and BPF_MAP_DELETE_ELEM.
+TEST_CASE("BPF_MAP_GET_NEXT_KEY", "[libbpf]")
```
Comment (saxena-anurag) [inline tests/unit/libbpf_test.cpp:1286]: Same as above: should we rename the test case?

**Review by saxena-anurag — APPROVED** [review]

**Review by saxena-anurag — APPROVED** [review]

**Pull request merged**